docs/quantum_gate.rst

```diff
@@ -1,4 +1,366 @@
+
+- directives; these gates execute neither in quantum nor in classical hardware; these look like gates but don't influence execution, e.g. the display gate
+
+
+Quantum gates can be subdivided in several ways from the number of operands they take; this becomes relevant when gates are mapped on the quantum computing platform:
+
+- one-qubit gates; quantum gates operating on one qubit
+
+- two-qubit gates; quantum gates operating on two qubits;
+  these are the main objective in mapping since two-qubit gates require their qubit operands to be connected in the hardware
+
+- multi-qubit gates; quantum gates operating (implicitly or explicitly) on more than two qubits;
+  these must be decomposed to one-qubit and two-qubit gates because more-qubit primitive gates are not supported by the quantum platform
+
+
+Quantum gates can also be subdivided seen from the state of a qubit:
+
+- preparation gates; (usually one-qubit) gates taking qubits in an undefined state and bringing them in a particular defined state
+
+- rotation gates; gates that perform unitary rotations on the state of the operand qubits;
+  examples are identity, x, rx(pi), cnot, swap, and toffoli.
+
+- measurement gates; gates that measure out the operand qubits, leaving them in a base state; the measurement result can be left in a classical register
+
+- scheduling gates; gates that only influence execution timing regarding the operand qubits;
+  they provide a cycle window for the qubit state to be operated upon before further use;
+  examples are the wait and barrier gates
+
+
+Particular classes of quantum gates can be further recognized:
+
+- primitive gates; quantum gates natively supported by instructions of the quantum computing platform
+
+- pauli gates; the Identity, X, Y and Z rotation gates
+
+- clifford gates;
+  a group of quantum gates (of 24 elements / equivalence classes)
+  each composed from a sequence of one or more rotations by a multiple of 90 degrees in one dimension (X, Y or Z)
+
+- default gates; quantum gates predefined by OpenQL
+
+- custom gates; quantum gates defined by the platform configuration file
+
+- composite gates; custom gates that are decomposed to their component gates when created
+
+- specialized gates; custom gates with a definition in the configuration file
+  that is specific for the particular qubit operands that are specified in it;
+  the semantic attributes of several specialized gates
+  with the same quantum operation but different qubit operands may differ
+  (in-line with the purpose of a gate being specialized)
+
+- parameterized gates; custom gates that are not specialized,
+  i.e. with a definition that is not specific for particular qubit operands;
+  all gates created (usually for different qubits) from the same parameterized gate in the platform configuration file
+  have the same semantic attributes
+
+
+
+.. _quantum_gate_attributes_in_the_internal_representation:
```

Comment:
It would be great to link these things to the API autodoc of PR #286 . But probably more something for a later point in time.

Reply:
Yes, but these are attributes of a gate in the internal representation,
and not necessarily names of parameters in a gate creating API.
I.e. we are free to rename the attributes systematically without affecting the API.
Having given them the same name now seems to lead to confusion.